packages/mobx-state-tree/src/types/complex-types/array.ts

```diff
@@ -333,11 +338,16 @@ function reconcileArrayChildren<TT>(
             // both are empty, end
             break
         } else if (!hasNewNode) {
-            // new one does not exists, old one dies
-            oldNode.die()
+            // new one does not exists
+            nothingChanged = false
             oldNodes.splice(i, 1)
+            if (oldNode instanceof ObjectNode) {
+                // since it is going to be returned by pop/splice/shift better create it before killing it
+                // so it doesn't end up in an undead state
+                oldNode.createObservableInstanceIfNeeded()
```

Comment:
here's the key, since the dehancer will be called again by mobx for this value (which will call .value, that will call createObservableInstance) we have to make sure it gets called BEFORE death to follow a proper lifecycle